### PR TITLE
ServiceBus sessions support

### PR DESCRIPTION
**Pull request by alrod**

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/MessagingProvider.cs [hunk 1]
```diff
@@ -108,5 +153,19 @@ private MessageSender GetOrAddMessageSender(string entityPath, string connection
...
             return _messageSenderCache.GetOrAdd(cacheKey, new MessageSender(connectionString, entityPath));
         }
+
+        private ClientEntity GetOrAddClientEntity(string entityPath, string connectionString)
+        {
+            string cacheKey = $"{entityPath}-{connectionString}";
+            string[] arr = entityPath.Split('/');
+            if (arr.Length > 1)
+            {
+                return _clientEntityCache.GetOrAdd(cacheKey, new SubscriptionClient(connectionString, arr[0], arr[2]));
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/MessagingProvider.cs]: Can you add a comment for what's happening here? I see you split it and then take [0] and [2] -- what about [1]? Maybe having a sample string that you're looking for would make it more clear.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusTriggerAttribute.cs [hunk 2]
```diff
@@ -83,6 +83,16 @@ public string SubscriptionName
...
 
+        /// <summary>
+        /// Gets or sets a value indicating whether the sessions are enabled.
+        /// </summary>
+        public bool IsSessionsEnabled { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Session Id
+        /// </summary>
+        public string SessionId { get; set; }
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusTriggerAttribute.cs]: Does this ever get used?

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Config/ServiceBusOptions.cs [hunk 3]
```diff
@@ -27,6 +27,10 @@ public ServiceBusOptions()
             {
                 MaxConcurrentCalls = 16
             };
+
+            SessionHandlerOptions = new SessionHandlerOptions(ExceptionReceivedHandler)
+            {
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Config/ServiceBusOptions.cs]: nit: remove unecessary {}

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Config/ServiceBusOptions.cs [hunk 4]
```diff
@@ -60,11 +70,24 @@ public string Format()
                 };
             }
 
+            JObject sessionHandlerOptions = null;
+            if (SessionHandlerOptions != null)
+            {
+                sessionHandlerOptions = new JObject
+                {
+                    { nameof(SessionHandlerOptions.AutoComplete), MessageHandlerOptions.AutoComplete },
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Config/ServiceBusOptions.cs]: So when not set, you're defaulting to the MessageHandlerOptions values? Why not just accept the default SB SessionHandlerOptions? That might be what users would want/expect?

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusEntityParam.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.ServiceBus;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Azure.WebJobs.ServiceBus
+{
+    internal class ServiceBusEntityParam
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusEntityParam.cs]: Recommend renaming this something like ServiceBusEntityInfo

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs [hunk 6]
```diff
@@ -46,8 +59,27 @@ public Task StartAsync(CancellationToken cancellationToken)
...
+                }
+                else
+                {
+                    SubscriptionClient subscriptionClient = _clientEntity as SubscriptionClient;
+                    subscriptionClient.RegisterSessionHandler(ProcessSessionMessageAsync, _serviceBusOptions.SessionHandlerOptions);
+                }
+            }
+            else
+            {
+                _receiver = _messagingProvider.CreateMessageReceiver(_entity.EntityPath, _serviceBusAccount.ConnectionString);
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs]: Technically we could be going through the new CreateClientEntity path for non-sessions right? We could create a Queue/Subscription Client and call RegisterMessageHandler. I get that we have a back compat issue here though.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusTriggerAttribute.cs (hunk 2 above)
Comment (mathewc) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusTriggerAttribute.cs]: I think this is a remnant from a previous design. I think we decided not to expose SessionId via attribute

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/SessionMessageProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+        /// <summary>
+        /// This method completes processing of the specified message, after the job function has been invoked.
+        /// </summary>
+        /// <remarks>
+        /// The message is completed by the ServiceBus SDK based on how the <see cref="MessageHandlerOptions.AutoComplete"/> option
+        /// is configured. E.g. if <see cref="MessageHandlerOptions.AutoComplete"/> is false, it is up to the job function to complete
+        /// the message.
+        /// </remarks>
+        /// <param name="session">The session to complete processing for.</param>
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/SessionMessageProcessor.cs]: nit: The session the message is part of

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Triggers/ServiceBusTriggerBinding.cs [hunk 8]
```diff
@@ -83,7 +70,7 @@ public async Task<ITriggerData> BindAsync(object value, ValueBindingContext cont
             }
 
             ITriggerData triggerData = await _argumentBinding.BindAsync(message, context);
-            IReadOnlyDictionary<string, object> bindingData = CreateBindingData(message, _listener?.Receiver, triggerData.BindingData);
+            IReadOnlyDictionary<string, object> bindingData = CreateBindingData(message, _listener?.Receiver, _listener?.ClientEntity, triggerData.BindingData); // SessionClient
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Triggers/ServiceBusTriggerBinding.cs]: remove this // SessionClient comment

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Triggers/ServiceBusTriggerBinding.cs [hunk 9]
```diff
@@ -126,6 +106,7 @@ public async Task<IListener> CreateListenerAsync(ListenerFactoryContext context)
             contract.Add("CorrelationId", typeof(string));
             contract.Add("UserProperties", typeof(IDictionary<string, object>));
             contract.Add("MessageReceiver", typeof(MessageReceiver));
+            contract.Add("ClientEntity", typeof(ClientEntity));
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Triggers/ServiceBusTriggerBinding.cs]: Users will have to cast in their code to the correct derived type (Queue/Subscription)?

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Triggers/ServiceBusTriggerBinding.cs [hunk 10]
```diff
@@ -185,16 +167,16 @@ private static void SafeAddValue(Action addValue)
 
         public ParameterDescriptor ToParameterDescriptor()
         {
-            string entityPath = _queueName != null ?
-                    _queueName : string.Format(CultureInfo.InvariantCulture, "{0}/Subscriptions/{1}", _topicName, _subscriptionName);
+            //string entityPath = _queueName != null ?
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Triggers/ServiceBusTriggerBinding.cs]: clean this up

File: test/Microsoft.Azure.WebJobs.Host.EndToEndTests/ServiceBusEndToEndTests.cs [hunk 11]
```diff
@@ -296,6 +296,12 @@ private async Task ServiceBusEndToEndInternal(IHost host = null)
                     "      \"MaxAutoRenewDuration\": \"00:05:00\",",
                     "      \"MaxConcurrentCalls\": 16",
                     "  }",
+                    "  \"SessionHandlerOptions\": {",
```
Comment (mathewc) [inline test/Microsoft.Azure.WebJobs.Host.EndToEndTests/ServiceBusEndToEndTests.cs]: You should add a full E2E test that exercises the new session support.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusEntityParam.cs (hunk 5 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusEntityParam.cs]: we already class ServiceBusEntity defined. Renamed to ServiceBusEntityInfo.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs (hunk 6 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs]: Technically yes. I left MessageReceiver.RegisterMessageHandler for back compatibility.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Triggers/ServiceBusTriggerBinding.cs (hunk 9 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Triggers/ServiceBusTriggerBinding.cs]: Yes, it was the idea.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Config/ServiceBusOptions.cs (hunk 4 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Config/ServiceBusOptions.cs]: It was copy/past mistake. Good catch.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/MessagingProvider.cs (hunk 1 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/MessagingProvider.cs]: Done.

File: test/Microsoft.Azure.WebJobs.Host.EndToEndTests/ServiceBusEndToEndTests.cs (hunk 11 above)
Comment (alrod) [reply test/Microsoft.Azure.WebJobs.Host.EndToEndTests/ServiceBusEndToEndTests.cs]: test/Microsoft.Azure.WebJobs.Host.EndToEndTests/ServiceBusSessionsEndToEndTests.cs

**Pull request merged**